homeassistant/components/owfs.py

```diff
@@ -0,0 +1,468 @@
+        self.hass = hass
+        self.dev = dev
+        self._events = None
+
+        name = config.get(CONF_NAME)
+        if name is None:
+            name = "{} {}".format(self.__class__.__name__, dev.id)
+        self._name = name
+        self._config = config
+
+        # save a ref to this device
+        try:
+            reg = dev.__objects
+        except AttributeError:
+            dev.__objects = reg = WeakValueDictionary()
+        attr = config.get(CONF_ATTRIBUTE)
+        if attr is None:
+            attr = "_default"
+        else:
+            self._attr = attr
+        reg[attr] = self
+
+    async def _init(self, without_config: bool):
+        """Async part of device initialization.
+        This code processes queued events and ensures that no more events
+        may be queued."""
+        events, self._events = self._events, None
+        for e in events:
+            await self.process_event(e)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def should_poll(self):
+        """overrides this Entity property.
+        OWFS doesn't require polling by HASS.
+        """
+        if not self._do_poll:
+            return False
+        t = time()
+        if self._next_poll > t:
+            return False
+        self._next_poll = t + self._do_poll
+        return True
+
+    @property
+    def unique_id(self):
+        """overrides this Entity property.
+        Return the OWFS device ID (considered unique).
+        """
+        res = self.dev.id
+        if self._attr is not None:
+            res += "_" + self._attr.replace('/', '_')
+        return res
+
+    @property
+    def available(self):
+        """overrides this Entity property.
+        Returns True if the device's bus address is known.
+        """
+        return self.dev.bus is not None
+
+    def __repr__(self):
+        rd = repr(self.dev)[1:]
+        if self.name and not self.name.endswith(self.dev.id):
+            rd = self.name + ' ' + rd
+        return "<%s:%s>" % (self.__class__.__name__, rd)
+
+    async def process_event(self, evt: "trio_owfs.event.DeviceEvent"):
+        """handle an event from OWFS.
+
+        If the event is an alarm, trigger a message.
+        """
+        from trio_owfs.event import DeviceAlarm, DeviceValue
+
+        if isinstance(evt, DeviceAlarm):
+            self.hass.bus.async_fire(
+                EVENT_OWFS_ALARM, {
+                    SERVICE_OWFS_ADDRESS: evt.device.id,
+                    SERVICE_OWFS_RESULTS: evt.device.results,
+                })
+        elif isinstance(evt, DeviceValue):
+            if self._attr is not None and evt.attribute == self._attr:
+                await self.async_update_ha_state()
+                self._val = evt.value
+
+    async def async_update(self):
+        if self.dev.bus is None:
+            return
+        self._val = await self.dev.attr_get(*self._attr.split('/'))
+
+    @property
+    def state(self):
```

Comment:
abstract base classes determine what `state` returns. It's therefore impossible that an entity mix-in can contain a state property.